File: rust/wxdragon-sys/cpp/src/activity_indicator.cpp
#include "../include/wxdragon.h" // Use correct path for wxdragon.h
#include <wx/activityindicator.h>
#include <wx/gdicmn.h> // For wxPoint, wxSize, wxDefaultPosition, wxDefaultSize
#include <wx/window.h> // For wxWindow parent type

WXD_EXPORTED wxd_ActivityIndicator_t *wxd_ActivityIndicator_Create(wxd_Window_t *parent, int id, int x, int y, int w, int h, long style) {
    wxWindow *p = (wxWindow *)parent;
    wxPoint pos = (x == -1 && y == -1) ? wxDefaultPosition : wxPoint(x, y);
    wxSize size = (w == -1 && h == -1) ? wxDefaultSize : wxSize(w, h);
    // wxActivityIndicator does not take a style parameter in its constructor.
    // The 'style' parameter from C API is currently ignored for this widget.
    // It could be used if there were relevant window styles to apply.
    wxActivityIndicator *indicator = new wxActivityIndicator(p, id, pos, size);
    return (wxd_ActivityIndicator_t *)indicator;
}

WXD_EXPORTED void wxd_ActivityIndicator_Start(wxd_ActivityIndicator_t *self) {
    wxActivityIndicator *indicator = (wxActivityIndicator *)self;
    if (!indicator) return;
    indicator->Start();
}

WXD_EXPORTED void wxd_ActivityIndicator_Stop(wxd_ActivityIndicator_t *self) {
    wxActivityIndicator *indicator = (wxActivityIndicator *)self;
    if (!indicator) return;
    indicator->Stop();
}

WXD_EXPORTED bool wxd_ActivityIndicator_IsRunning(wxd_ActivityIndicator_t *self) {
    wxActivityIndicator *indicator = (wxActivityIndicator *)self;
    if (!indicator) return false;
    return indicator->IsRunning();
} 

File: rust/wxdragon/src/widgets/activity_indicator.rs
use crate::geometry::{Point, Size};
use crate::event::WxEvtHandler;
use crate::id::Id;
use crate::implement_widget_traits;
use crate::widget_builder;
use crate::widget_style_enum;
use crate::window::{Window, WxWidget};
use wxdragon_sys as ffi;

// Define a proper style enum for ActivityIndicator
widget_style_enum!(
    name: ActivityIndicatorStyle,
    doc: "Style flags for ActivityIndicator control.",
    variants: {
        Default: 0, "Default style."
    },
    default_variant: Default
);

/// Represents a `wxActivityIndicator`, an animated control that shows 
/// an animation to indicate a long-running process is occurring.
#[derive(Clone)]
pub struct ActivityIndicator {
    window: Window,
}

impl ActivityIndicator {
    /// Creates a new `ActivityIndicatorBuilder` for constructing an activity indicator.
    pub fn builder(parent: &dyn WxWidget) -> ActivityIndicatorBuilder {
        ActivityIndicatorBuilder::new(parent)
    }

    /// Low-level constructor used by the builder.
    fn new_impl(
        parent_ptr: *mut ffi::wxd_Window_t,
        id: Id,
        pos: Point,
        size: Size,
        style: i64,
    ) -> Self {
        assert!(!parent_ptr.is_null(), "ActivityIndicator requires a parent");
        
        let ptr = unsafe {
            ffi::wxd_ActivityIndicator_Create(
                parent_ptr,
                id,
                pos.x,
                pos.y,
                size.width,
                size.height,
                style,
            )
        };
        
        if ptr.is_null() {
            panic!("Failed to create wxActivityIndicator");
        }
        
        unsafe {
            let window = Window::from_ptr(ptr as *mut ffi::wxd_Window_t);
            ActivityIndicator { window }
        }
    }

    /// Start the animation.
    pub fn start(&self) {
        unsafe { ffi::wxd_ActivityIndicator_Start(self.window.as_ptr() as *mut _) }
    }

    /// Stop the animation.
    pub fn stop(&self) {
        unsafe { ffi::wxd_ActivityIndicator_Stop(self.window.as_ptr() as *mut _) }
    }

    /// Check if the animation is currently running.
    pub fn is_running(&self) -> bool {
        unsafe { ffi::wxd_ActivityIndicator_IsRunning(self.window.as_ptr() as *mut _) }
    }
}

// Use the widget_builder macro for ActivityIndicator
widget_builder!(
    name: ActivityIndicator,
    parent_type: &'a dyn WxWidget,
    style_type: ActivityIndicatorStyle,
    fields: {},
    build_impl: |slf| {
        ActivityIndicator::new_impl(
            slf.parent.handle_ptr(),
            slf.id,
            slf.pos,
            slf.size,
            slf.style.bits(),
        )
    }
);

// Apply common trait implementations for this widget
implement_widget_traits!(ActivityIndicator, window);
